File: cart/Cartridge4K.java
//============================================================================
//
//   SSSS    tt          lll  lll
//  SS  SS   tt           ll   ll
//  SS     tttttt  eeee   ll   ll   aaaa
//   SSSS    tt   ee  ee  ll   ll      aa
//      SS   tt   eeeeee  ll   ll   aaaaa  --  "An Atari 2600 VCS Emulator"
//  SS  SS   tt   ee      ll   ll  aa  aa
//   SSSS     ttt  eeeee llll llll  aaaaa
//
// Copyright (c) 1995-2007 by Bradford W. Mott and the Stella team
//
// See the file "license" for information on usage and redistribution of
// this file, and for a DISCLAIMER OF ALL WARRANTIES.
//
// $Id: Cartridge4K.java,v 1.2 2007/08/12 04:51:29 mauvila Exp $
//============================================================================
package jstella.cart;


import jstella.j6507.*;
import jstella.j6507.IfcSystem;

public class Cartridge4K extends Cartridge {
    private final static long serialVersionUID = -3240308689643904765L;
    
    private final static int CARTRIDGE_MASK_VALUE=0x0FFF;
    private final static int CARTRIDGE_SIZE=4096;
    private final static int CARTRIDGE_BANK_COUNT=1;
    
    
    
    
    // The 4K ROM image for the cartridge
    int[] myImage=new int[CARTRIDGE_SIZE];
    
    
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    public Cartridge4K(int[] image) {
        
        myImage=copyImage(image);
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    public String name() {
        return "Cartridge4K";
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    public void reset() {
    }
    
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    
    
    public void install(jstella.core.JSSystem system) {
        mySystem = system;        
        addDirectPeekAccess(0x1000, 0x2000, myImage, CARTRIDGE_MASK_VALUE);  // Map ROM image into the system
        
    }
    
    
    
    
    
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    public int peek(int address) {
        return myImage[address & CARTRIDGE_MASK_VALUE];
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    public void poke(int aC, int aB) {
        // This is ROM so poking has no effect :-)
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    protected void setCurrentBank(int bank) {
        // Doesn't support bankswitching
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    protected int getCurrentBank() {
        // Doesn't support bankswitching
        return 0;
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    protected int bankCount() {
        return CARTRIDGE_BANK_COUNT;
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    public boolean patch(int address,int value) {
        myImage[address & CARTRIDGE_MASK_VALUE] = value;
        return true;
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    public int[] getImage() {
        
        return myImage;
    }
    
    
    
}

File: cart/Cartridge4K.js
//============================================================================
//
//   SSSS    tt          lll  lll
//  SS  SS   tt           ll   ll
//  SS     tttttt  eeee   ll   ll   aaaa
//   SSSS    tt   ee  ee  ll   ll      aa
//      SS   tt   eeeeee  ll   ll   aaaaa  --  "An Atari 2600 VCS Emulator"
//  SS  SS   tt   ee      ll   ll  aa  aa
//   SSSS     ttt  eeeee llll llll  aaaaa
//
// Copyright (c) 1995-2007 by Bradford W. Mott and the Stella team
//
// See the file "license" for information on usage and redistribution of
// this file, and for a DISCLAIMER OF ALL WARRANTIES.
//
// $Id: Cartridge4K.java,v 1.2 2007/08/12 04:51:29 mauvila Exp $
//============================================================================
function Cartridge4K (image){
    this.serialVersionUID = "-3240308689643904765L";
    
    this.CARTRIDGE_MASK_VALUE=0x0FFF;
    this.CARTRIDGE_SIZE=4096;
    this.CARTRIDGE_BANK_COUNT=1;
    
    
    
    
    // The 4K ROM image for the cartridge
    this.myImage=[];
    
    
		//이하 원래 상속 

this.setConsole = function(aConsole) { myConsole=aConsole; };

  	this.systemCyclesReset = function() {    };
    
		this.getMD5 = function(){ return this. myMD5; };
    this.setMD5 = function(aMD5){ this.myMD5=aMD5; };
    
		this.copyImage = function(aSourceImage)
    {
        var zReturn = [];
				var length = aSourceImage.length;
        for (var i=0; i<length; i++)
        {
            zReturn[i]=aSourceImage[i];
        }//end : for i loop
        return zReturn;
    };
    

    this.addIndirectAccess = function(aStartAddress, aEndAddress)
    {
        
        for(var zAddress = (aStartAddress & ~PAGE_MASK); zAddress < (aEndAddress & ~PAGE_MASK); zAddress += PAGE_SIZE)
        {
            this.mySystem.setPageAccess(zAddress >> PAGE_SHIFT, (new PageAccess()).createIndirectAccess(this));
        } 
    };
    
    this.addDirectPeekAccess = function(aStartAddress, aEndAddress, aMemory, aBaseAddressMask, aBaseAddressOffset)
     {
			 	if(arguments.length == 4)
	         this.addDirectPeekAccess(aStartAddress, aEndAddress, aMemory, aBaseAddressMask, 0);
				else  {
		        for(var zAddress = (aStartAddress & ~PAGE_MASK); zAddress < (aEndAddress & ~PAGE_MASK); zAddress += PAGE_SIZE)
    	    {
      	      this.mySystem.setPageAccess(zAddress >> PAGE_SHIFT, (new PageAccess()).createDirectPeekAccess(this, aMemory, aBaseAddressOffset + (zAddress & aBaseAddressMask)));
        	}  
				}
     };
    
    this.addDirectPokeAccess = function(aStartAddress, aEndAddress, aMemory, aBaseAddressMask, aBaseAddressOffset)
    {
        for(var zAddress = (aStartAddress & ~PAGE_MASK); zAddress < (aEndAddress & ~PAGE_MASK); zAddress += PAGE_SIZE)
        {
            this.mySystem.setPageAccess(zAddress >> PAGE_SHIFT, (new PageAccess()).createDirectPokeAccess(this, aMemory, aBaseAddressOffset + (zAddress & aBaseAddressMask)));
        } 
    }
    
    this.addDirectPokeAccess = function(aStartAddress, aEndAddress,aMemory, aBaseAddressMask)
     {
         this.addDirectPokeAccess(aStartAddress, aEndAddress, aMemory, aBaseAddressMask, 0);
     }
    //상속 끝

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.myImage=this.copyImage(image); //Constructor
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.name = function() {
        return "Cartridge4K";
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.reset = function() {
    }
    
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    
    
    this.install = function(system) { //JSSystem?
        this.mySystem = system;        
        this.addDirectPeekAccess(0x1000, 0x2000, this.myImage, this.CARTRIDGE_MASK_VALUE);  // Map ROM image into the system
        
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.peek = function(address) {
        return myImage[address & this.CARTRIDGE_MASK_VALUE];
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.poke = function(aC, aB) {
        // This is ROM so poking has no effect :-)
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.setCurrentBank = function(bank) {
        // Doesn't support bankswitching
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.getCurrentBank = function() {
        // Doesn't support bankswitching
        return 0;
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.bankCount = function() {
        return this.CARTRIDGE_BANK_COUNT;
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.patch = function(address, value) {
        myImage[address & this.CARTRIDGE_MASK_VALUE] = value;
        return true;
    }
    
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    this.getImage = function() {
        
        return myImage;
    }
    
    
    
}
